Dictionary and wikipedia enhancements (#2393)

poire-z opened the pull request:
Feedback wanted on ReaderDictionary:cleanSelection():
It looks a bit heavy and redundant, but all that was needed to clean some french book selections (well, may be it could work with 1 or 2 gsub less, but i was tired testing :)
Dunno if there may be some side effects with other languages.

Selection of multiple words in TextBoxWidget:onHoldReleaseText() seems fine on western languanges, but needs to be checked on chinese or japanese texts (should be at least not worse on single word than the old Hold method).

houqp commented:
Thanks for the patch. It will take me some time to go through all of it, hopefully I can find some free time tomorrow :)

poire-z commented on frontend/ui/widget/dictquicklookup.lua:
```diff
@@ -106,7 +143,7 @@ function DictQuickLookup:update()
         h = Screen:getHeight(),
     }
-    if self.word_box then
+    if self.word_box and not not self.fullpage then
```
Obviously, i had been too negative there :) I'll remove one of these `not` at next changes request (dunno if one started reviewing this, and if one can diff commits, if i commit a fix for that now)

houqp replied:
ha, good catch.

poire-z commented on frontend/apps/reader/modules/readerdictionary.lua:
```diff
@@ -13,6 +13,7 @@ local T = require("ffi/util").template
 local ReaderDictionary = InputContainer:new{
     data_dir = nil,
     dict_window_list = {},
+    lookup_msg = _("Dictionary lookup of :\n“%1” ...")
```
Not sure if that's correct english, or if there's a better wording (but i'd like to keep what was selected on a line by itself)

Frenzie replied:
It's grammatical, but you're probably looking for something like dictionary result or definition instead? NB I haven't really checked when this would show up or anything.

poire-z replied:
It's displayed as an InfoMessage while lookup is in progress, and discarded when done. I'm used to see "Looking up <text>...", but my "(Dictionnary|Wikipedia) lookup of <text>" sounds strange now.

Frenzie replied:
Yeah, I'd probably say something like "(looking up|searching for) (?:definition of)"

houqp replied:
I am also used to `Looking up ...`

poire-z replied:
Yes, but where to put `Dictionary` or `Wikipedia`, and the search term(s) ?
I'd like them to strike in the few ms/seconds the InfoMessage is visible.
`Looking up "19th century in literature" on Wikipedia...` or
`Looking up 'maintainers' in Dictionary...`
with a possible wrap in the middle, make it hard to catch the selected text.

Frenzie replied:
Alright, call it something like "Searching Wikipedia|dictionary for:" instead of trying to noun the phrasal verb *look up* and call it a day. =P

poire-z replied:
Nice. Done. Day !

houqp commented on frontend/ui/widget/dictquicklookup.lua:
```diff
@@ -185,25 +254,36 @@ function DictQuickLookup:update()
                 {
-                    text = _("Add Note"),
-                    enabled = false,
+                    text = self.wiki and ( #self.wiki_languages > 1 and self.wiki_languages[1].." > "..self.wiki_languages[2] or self.wiki_languages[1] ) or "-",
```
again, better not to abuse `foo and x or y` here for complex conditions like this.

houqp commented on frontend/ui/widget/dictquicklookup.lua:
```diff
@@ -331,6 +420,19 @@ function DictQuickLookup:changeDictionary(index)
+    end
+    -- add queried word to 1st result's definition, so we can see
+    -- what was the selected text and if we selected wrong
+    if index == 1 and not self.fullpage then
```
The `not self.fullpage` branch here can be folded into previous else branch so you are not checking it twice.

houqp commented on frontend/ui/widget/textboxwidget.lua:
```diff
@@ -112,7 +121,14 @@ function TextBoxWidget:_splitCharWidthList()
+            -- a colon, and it better not be wrapped there).
+            local next_c = idx+1 <= size and self.char_width_list[idx+1].char or "willnotbefound"
+            -- was : if util.isSplitable(c) then
+            if c == " " and not non_splittable_space_tailers:find(next_c, 1, true) then
```
this breaks line breaking for CJK. If you look at the definition of `util.isSplitable`. I suggest you pass the next_c as a optional second argument to `util.isSplitable` and move the logic there. Also it's better to set `next_c` as `nil` if `c` is at the end of the line.

houqp commented on frontend/ui/widget/textboxwidget.lua:
```diff
@@ -144,6 +163,15 @@ function TextBoxWidget:_splitCharWidthList()
+            if idx <= size and self.char_width_list[idx].char == " " then
+                -- It's probably not needed to append it to last line
+                -- (not sure if that's ok with cursor stuff)
+                -- self.vertical_string_list[ln].text = self.vertical_string_list[ln].text .. " "
```
why leaving out this commented line here?

poire-z replied:
Previously, you were either appending this space to this line, or prepending it to next line.
With my aditional checks, we'll be there with this space that we split/wrap lines on.
I think it makes no difference appending this space to current line, or ignoring it, except for the following:
- if we append it, should we increase `self.vertical_string_list[ln].width` with this space's size for the sake of coherency ? even if we splitted there so width does not exceed a limit.
- if we append it, we will render it, and this space may eat some bit of the scroll bar (depends on font, and there is there this HorizontalSpan that will make that ok)
- again, i dont know about cursor stuff, I don't know where there is an editable textboxwidget to test with

Also, when meeting a hard newline, you don't add it to any vertical_string_list[ln].text, you just use it to do the split. I think it's the same for this space at this point in the code.
So, i think just saying that in a comment is enough. Should i add what i just wrote to that comment, if you agree with it ?

houqp replied:
I think it's fine to just ignore spaces on line break. Cursor uses `self.vertical_string_list` too so it should be fine. To enable cursor, just set `TextBoxWidget.editable` to `true`. I would just write a comment says `discard spaces on line break boundary`.

houqp commented on frontend/ui/widget/textboxwidget.lua:
```diff
@@ -333,4 +361,121 @@ function TextBoxWidget:onHoldWord(callback, ges)
+    end
+
+    local selected_text = table.concat(self.charlist, "", sel_start_idx, sel_end_idx)
+    DEBUG("onHoldReleaseText (duration: "..hold_duration..") : "..
```
Don't do string concatenation for debug call. The `DEBUG` call should be a no op for non-debug mode. String concatenation will create a new string which introduces runtime penalty. You can do `DEBUG("onHoldReleaseText (duration: ", hold_duration, "): ", sel_start_idx, ">", sel_end_idx, " = ", selected_text)` instead.

houqp commented on frontend/ui/widget/textboxwidget.lua:
```diff
@@ -333,4 +361,121 @@ function TextBoxWidget:onHoldWord(callback, ges)
+    end
+
+    -- similar code to find start or end is in _findWordEdge() helper
+    local sel_start_idx = self:_findWordEdge(x0, y0, "start")
```
Better use module global int flags for things like this, it's much more efficient for comparison:

```lua
local FIND_START = 1
local FIND_END = 2
```

poire-z commented on frontend/apps/reader/modules/readerdictionary.lua:
```diff
@@ -64,20 +65,90 @@ local function tidy_markup(results)
     return results
 end
 
+function ReaderDictionary:cleanSelection(text, is_single_word)
```
I just made a mistake in ReaderDictionary:stardictLookup() where it should be called with `false`. I assume stardict needs a word, without spaces, unlike wikipedia where multiple words are ok. This `is_single_word` is there so we can do different cleaning for each cases.

poire-z replied:
I finally got rid of is_single_word. It was there because in readerdictionary, there was a call to stripePunctuation, that wasn't there in readerwikipedia. So i thought i would do as it was done, but stripePunctuation is as well useful in both cases.

poire-z commented on frontend/apps/reader/modules/readerdictionary.lua:
```diff
@@ -64,20 +65,90 @@ local function tidy_markup(results)
+function ReaderDictionary:cleanSelection(text, is_single_word)
+    -- Will be used by ReaderWikipedia too (with is_single_word false)
+    -- We do multiple times the same replacements, which is most of the time overkill,
+    -- but sometimes provices better cleaning
```
Can i just put there in comments some sample selections crengine has returned ?
Don't ask me to do some unit test please :| I dont think it's that critical to be consistent here. It's a mixture of crengine word detection algorithm (which may be greedy), and real life epub editors' various typography rules (using various kind of dash, unbreakable spaces, or not)

houqp replied:
Yes, adding samples in comment should be fine for now. I can help convert them to uni tests later, it's actually very easy to add one. I am more concerned whether doing multiple passes of replacements is the correct approach here, which can be verified with examples. In theory, one pass should be enough to extract the needed word with lpeg.

Yes, it's fine to have language specific rules here.

poire-z replied:
I removed quite a bit of these gsub. Tested on some extrem cases, and it works as fine.
I don't think we have to worry much about time spend in these gsubs. Compared to the time it will take to do the stardict or wikipedia lookup, it's nothing. And we do that only when the user asks for a lookup.
Anyway, i think it looks cleaner, and clear alongside my samples in comments.

poire-z commented on frontend/ui/widget/dictquicklookup.lua:
```diff
@@ -185,25 +254,36 @@ function DictQuickLookup:update()
                 {
-                    text = _("Wikipedia"),
-                    enabled = not self.wiki,
+                    text = self.wiki and _("Full page") or _("Wikipedia"),
```
I'll do that.
(But no, here, i was using self.wiki to distinguish wiki from dict : if wiki, this button is "Full Page" - if dict, this button is "Wikipedia" like before. We're already in the "not fullpage" buttons branch - but ok, will be easier to read without these x and y or z)

poire-z commented:
One unit test failed:
```
util module should split text to line - unicode
spec/front/unit/util_spec.lua:95: Expected objects to be the same.

Passed in: <= that is was is expected
(table) {
 *[1] = 'Pójde,'
  [2] = ' '
  [3] = 'chmurno '
  [4] = 'glück '
  [5] = 'schließen '
  [6] = '~Jtstí '
  [7] = 'ne~Ztstí.'
  [8] = ' '
  [9] = 'Uñas '
  [10] = 'gavilán' }
Expected:     <= that is what we get with my modifications
(table) {
 *[1] = 'Pójde, '
  [2] = 'chmurno '
  [3] = 'glück '
  [4] = 'schließen '
  [5] = '~Jtstí '
  [6] = 'ne~Ztstí. '
  [7] = 'Uñas '
  [8] = 'gavilán' }
```
I think that what we get now is the expected better decision. So, if you agree, i'll update that unit spec too.
(I have a small other thing to update (next_c forgotten in one of my 2 calls to isSplittable), so i was going to update a commit anyway)

poire-z commented:
Anyway, I did most of the suggested changes:
- a single word with the selected text (whether it's a query or a full page title)
- booleans is_wiki , get_fullpage (when we want to get a fullpage), is_fullpage (when the result or window is a fullpage result)
- less is_wiki and x or y, except for when it would be uglier to push the logic away (including some cases where you suggested a change)
- moved my non_splittable_space_tailers logic into util.isSplitable
- fixed a bug when selecting chinese chars (i always got 2 chars, thought it was some eastern alchemy that i would never get, and let that to native readers - but it was actually western lazyness of mine : i always got a space after my selection, and i just forgot to add a `idx-1` ...)

Btw, one question, something i haven't found out by myself : when/what calls the :init() method ?
Also, i see that local DictQuickLookup = InputContainer:new{ ) contains lots of Screen:scaleBySize() (i put mine there too now) : is it for economy reasons because this is executed only once ?

houqp commented:
Thanks for the update, all looks good to me now :) Also thanks for adding the unit test ;)

`init` is called inside the `new` function, see `function Widget:new` in `frontend/ui/widget/widget.lua`.

With regards to scaleBySize, it's added there to make the sizing portable between devices with different DPIs. The same pixel unit will be rendered small on devices with higher DPI, which makes some of the UI component too small to be usable for those devices (often newer models).
